### ISSUE TITLE
"Unknown operation for batch reader" after upgrading to 0.9.1 from 0.7

### ISSUE THREAD

**Issue by wichtounet** [deeplearning4j/deeplearning4j#4189]
Hi,

I'm currently upgrading my benchmark test bed with DL4J from 0.7 to 0.9.1 and I'm having an issue on my ImageNet benchmark. It was working well before, but now there is a runtime exception: 

```
Exception in thread "ADSI prefetch thread" java.lang.RuntimeException: java.lang.UnsupportedOperationException: Unknown operation for batch reader
	at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:442)
Caused by: java.lang.UnsupportedOperationException: Unknown operation for batch reader
	at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertWritablesBatched(RecordReaderMultiDataSetIterator.java:301)
	at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertFeaturesOrLabels(RecordReaderMultiDataSetIterator.java:267)
	at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.nextMultiDataSet(RecordReaderMultiDataSetIterator.java:234)
	at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.next(RecordReaderMultiDataSetIterator.java:177)
	at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:306)
	at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:393)
	at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:51)
	at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:417)
```

I guess the problem comes from the data set iterator, but I don't know what to change. Here is the full code: 

```java
public class experiment6 {
    private static final Logger log = LoggerFactory.getLogger(experiment6.class);

    public static void main(String[] args) throws Exception {
        int nChannels = 3; // Number of input channels
        int outputNum = 1000; // The number of possible outcomes
        int batchSize = 64; // Test batch size
        int numEpochs = 10; // Number of training epochs

        File parentDir = new File("/data/datasets/imagenet_resized/train/");
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();
        ImageRecordReader recordReader = new ImageRecordReader(256,256,nChannels,labelMaker);
        recordReader.initialize(new FileSplit(parentDir));

        DataSetIterator dataIter = new RecordReaderDataSetIterator(recordReader,batchSize);

        log.info("Build model....");
        MultiLayerConfiguration.Builder builder = new NeuralNetConfiguration.Builder()
                .trainingWorkspaceMode(WorkspaceMode.SINGLE)
                .iterations(1)
                .regularization(false)
                .learningRate(0.01)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.SGD)
                .momentum(0.9)
                .list()
                .layer(0, new ConvolutionLayer.Builder(3, 3).stride(1, 1).padding(1, 1).nIn(nChannels).nOut(16).activation("relu").build())
                .layer(1, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX).kernelSize(2,2).stride(2,2).build())
                .layer(2, new ConvolutionLayer.Builder(3, 3).stride(1, 1).padding(1, 1).nOut(16).activation("relu").build())
                .layer(3, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX).kernelSize(2,2).stride(2,2).build())
                .layer(4, new ConvolutionLayer.Builder(3, 3).stride(1, 1).padding(1, 1).nOut(32).activation("relu").build())
                .layer(5, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX).kernelSize(2,2).stride(2,2).build())
                .layer(6, new ConvolutionLayer.Builder(3, 3).stride(1, 1).padding(1, 1).nOut(32).activation("relu").build())
                .layer(7, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX).kernelSize(2,2).stride(2,2).build())
                .layer(8, new ConvolutionLayer.Builder(3, 3).stride(1, 1).padding(1, 1).nOut(32).activation("relu").build())
                .layer(9, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX).kernelSize(2,2).stride(2,2).build())
                .layer(10, new DenseLayer.Builder().activation("relu").nOut(2048).build())
                .layer(11, new OutputLayer.Builder(LossFunctions.LossFunction.MCXENT)
                        .nOut(outputNum)
                        .activation("softmax")
                        .build())
                .setInputType(InputType.convolutionalFlat(256,256,nChannels)) //See note below
                .backprop(true).pretrain(false);

        MultiLayerConfiguration conf = builder.build();
        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.setListeners(new ScoreIterationListener(1), new PerformanceListener(1));
        model.init();

        log.info("Train model....");
        for( int i=0; i<numEpochs; i++ ){
            log.info("Epoch " + i);
            model.fit(dataIter);
        }
    }
}
```

(Also available online: https://github.com/wichtounet/frameworks/blob/master/dl4j/src/main/java/wicht/experiment6.java)

How can I make this works with DL4J 0.9.1 ? 

Thanks

**Comment by saudet** [comment]
You forgot to give an index for the label and the number of labels, something like this:
```java
DataSetIterator dataIter = new RecordReaderDataSetIterator(recordReader, batchSize, 1, outputNum);
```

**Comment by AlexDBlack** [comment]
Right, just specify the label index + number of labels. Closing this.

**Comment by wichtounet** [comment]
Thanks, that worked :+1:
